<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Привет. Оформление комментариев по работе сохраняется. Только обозначим, что это третья итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Привет еще раз. Спасибо, что доделал работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>

Привет! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Меня зовут Слепцов Артем и я буду проверять твой проект) Ты проделал большую работу над проектом, однако его еще можно улучшить. Будет здорово, если ты, надеюсь, не против, если я буду на ты, будешь отвечать на комментарии и участвовать в диалоге. 

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках (если фон комментария <font color=green>зелёный</font> - всё сделано правильно. Рекомендации укажу таким же цветом; <font color=orange>оранжевый</font> - некритичные замечания. Если таких замечаний меньше трех - проект может быть принят без их отработки; <font color=red>красный</font> - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<h2> Комментарий ревьюера</h2>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе следует приводить краткий план того, что надо сделать(если этот план имеется), а также информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут. Также название работы лучше отображать в отдельной ячейке и делать крупный шрифт (заголовок). Так работа выглядит презентабельно. Подробнее смотри [методичку](https://www.notion.so/35d9c2f32ffa42f3b17010737aedca4b) по оформлению проектов.

</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
# загружаю таблицу, смотрю шапку
import pandas as pd
import numpy as np
import pprint

from pymystem3 import Mystem
from collections import Counter
df = pd.read_csv('/datasets/data.csv')
df.head(100)

children  days_employed  dob_years education  education_id  \
0          1   -8437.673028         42    высшее             0   
1          1   -4024.803754         36   среднее             1   
2          0   -5623.422610         33   Среднее             1   
3          3   -4124.747207         32   среднее             1   
4          0  340266.072047         53   среднее             1   
..       ...            ...        ...       ...           ...   
95         0    -541.832241         51    высшее             0   
96         0            NaN         44   СРЕДНЕЕ             1   
97         0            NaN         47    высшее             0   
98         0  364906.205736         54    высшее             0   
99         0  346541.618895          0   Среднее             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
..               ...               ...    ...         ...   ...   
95   женат / замужем                 0      F   сотрудник     0   
96   женат / замужем                 0      F   сотрудник     0   
97   женат / замужем                 0      F   сотрудник     0   
98   женат / замужем                 0      F   пенсионер     0   
99   женат / замужем                 0      F   пенсионер     0   

     total_income                     purpose  
0   253875.639453               покупка жилья  
1   112080.014102     приобретение автомобиля  
2   145885.952297               покупка жилья  
3   267628.550329  дополнительное образование  
4   158616.077870             сыграть свадьбу  
..            ...                         ...  
95   94187.878082                  автомобиль  
96            NaN  покупка жилой недвижимости  
97            NaN      профильное образование  
98  199707.298524     покупка жилья для сдачи  
99   71291.522491                  автомобиль  

[100 rows x 12 columns]

In [2]:
# смотрю обущую информацию
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
#ищу количество пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [4]:
#смотрю отклонения
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Вывод(шаг 1):
a) У нас таблица размером 12 на 21525
b) Есть пропуски в столбцах 'общий трудовой стаж в днях' и 'ежемесячный доход'. Скорее всего это связано, ведь без стажа нет дохода. 
c) Столбец days_employed имеет тип данных float64 и отрицательные значения. Нужно исправить. И для наглядности его указывать в годах.
d) В столбце purpose указывается цель кредита. Она может быть одна у многих, но сформулирована у всех по-разному.
e) Не указана валюта. (хотя она здесь не очень то нужна, мне кажется, потому что урвень жизни это понятие относительное. При условии что все указывают в одной валюте) 
f) Разный регистр в столбце 'Образование'
g) Минимальный возраст - 0, это ошибка.
h) Количество детей - -1 и 20, это явно ошибка.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

У меня имеется несколько комментариев по данному шагу:
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен. Здорово, что ты обнаружил ошибки в данных.

</div>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [5]:
#смотрю пропуски
display(df[df['total_income'].isnull()]) 

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [18]:
#группирую пропуски по профессии и смотрю медианные значения по каждой профессии
df_income_type = df.groupby('income_type').median()
display(df_income_type)

children  days_employed  dob_years  education_id  \
income_type                                                         
безработный           0.5  366413.652744       38.0           0.5   
в декрете             2.0   -3296.759962       39.0           1.0   
госслужащий           0.0   -2689.368353       40.0           1.0   
компаньон             0.0   -1547.382223       39.0           1.0   
пенсионер             0.0  365213.306266       60.0           1.0   
предприниматель       0.0    -520.848083       42.5           0.0   
сотрудник             0.0   -1360.363902       39.0           1.0   
студент               0.0    -578.751554       22.0           0.0   

                 family_status_id  debt   total_income  
income_type                                             
безработный                   0.5   0.5  131339.751676  
в декрете                     0.0   1.0   53829.130729  
госслужащий                   0.0   0.0  150447.935283  
компаньон                     0.0   0.0  172357.950966  
пенсионер                     0.0   0.0  118514.486412  
предприниматель               0.5   0.0  499163.144947  
сотрудник                     0.0   0.0  142594.396847  
студент                       4.0   0.0   98201.625314

In [19]:
#заменяю пропуск по стажу на медианные значения по каждой профессии
df.loc[df['income_type']=='безработный', ['days_employed']] = df.loc[df['income_type']=='безработный', 'days_employed'].fillna(df_income_type.loc['безработный', 'days_employed'])
df.loc[df['income_type']=='в декрете', ['days_employed']] = df.loc[df['income_type']=='в декрете', 'days_employed'].fillna(df_income_type.loc['в декрете', 'days_employed'])
df.loc[df['income_type']=='госслужащий', ['days_employed']] = df.loc[df['income_type']=='госслужащий', 'days_employed'].fillna(df_income_type.loc['госслужащий', 'days_employed'])
df.loc[df['income_type']=='компаньон', ['days_employed']] = df.loc[df['income_type']=='компаньон', 'days_employed'].fillna(df_income_type.loc['компаньон', 'days_employed'])
df.loc[df['income_type']=='пенсионер', ['days_employed']] = df.loc[df['income_type']=='пенсионер', 'days_employed'].fillna(df_income_type.loc['пенсионер', 'days_employed'])
df.loc[df['income_type']=='сотрудник', ['days_employed']] = df.loc[df['income_type']=='сотрудник', 'days_employed'].fillna(df_income_type.loc['сотрудник', 'days_employed'])
df.loc[df['income_type']=='студент', ['days_employed']] = df.loc[df['income_type']=='студент', 'days_employed'].fillna(df_income_type.loc['студент', 'days_employed'])






In [24]:
#делаю тоже самое с доходами
df.loc[df['income_type']=='безработный', ['total_income']] = df.loc[df['income_type']=='безработный', 'total_income'].fillna(df_income_type.loc['безработный', 'total_income'])
df.loc[df['income_type']=='в декрете', ['total_income']] = df.loc[df['income_type']=='в декрете', 'total_income'].fillna(df_income_type.loc['в декрете', 'total_income'])
df.loc[df['income_type']=='госслужащий', ['total_income']] = df.loc[df['income_type']=='госслужащий', 'total_income'].fillna(df_income_type.loc['госслужащий', 'total_income'])
df.loc[df['income_type']=='компаньон', ['total_income']] = df.loc[df['income_type']=='компаньон', 'total_income'].fillna(df_income_type.loc['компаньон', 'total_income'])
df.loc[df['income_type']=='пенсионер', ['total_income']] = df.loc[df['income_type']=='пенсионер', 'total_income'].fillna(df_income_type.loc['пенсионер', 'total_income'])
df.loc[df['income_type']=='сотрудник', ['total_income']] = df.loc[df['income_type']=='сотрудник', 'total_income'].fillna(df_income_type.loc['сотрудник', 'total_income'])
df.loc[df['income_type']=='студент', ['total_income']] = df.loc[df['income_type']=='студент', 'total_income'].fillna(df_income_type.loc['студент', 'total_income'])


In [25]:
display(df)

children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21524 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21524 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


пропусков больше нет

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Обработать пропуски таким образом мы не можем. Данные у нас о людях, уже получивших кредиты. Лучше использовать заполнение пропусков на основе группировки данных по одному из параметров в нашей таблице. Подумай, какой параметр будет больше всего влиять на доход. От чего больше всего зависит доход? Проведи группировку по нему и на его основании заполни пропуски. 

</div>

Я сгруппировал данные по профессии так как думаю, что доход зависит именно от нее. После этого я нашел медианные значения по каждой профессии и подставил их вместо пропусков. Пропусков больше нет. Я выбрал медиану так как в этом случае подставляемое значение избегает экстремальных значений.


<div class="alert alert-warning">
<h2> Комментарий ревьюера 2</h2>

По логике ты пропуски обработал правильно. Конечно, тип занятости сильно определяет доход. Только процесс обработки пропусков следует автоматизировать например, через цикл по уникальным типам занятости. 

</div>

### Замена типа данных

In [8]:
#меняю тип данных с float на int в days_employed
df['days_employed'] = pd.to_numeric(df['days_employed']).astype(int)

In [9]:
#делю количество дней на 365 для получения стажа в года
df['days_employed'] = df['days_employed'] / 365

In [10]:
#делаю количество дней положительными 
df['days_employed'] = df['days_employed'].abs()

In [11]:
#меняю тип данных с float на int в days_employed
#я использую именно этот способ, потому что он более универсальный 
df['days_employed'] = pd.to_numeric(df['days_employed']).astype(int)

In [12]:
#переименную столбец days_employed в years_employed для понимания
df = df.rename(columns={'days_employed': 'years_employed'})

In [13]:
#теперь гораздо проще понять данные
df.describe()

children  years_employed     dob_years  education_id  \
count  21525.000000    21525.000000  21525.000000  21525.000000   
mean       0.538908      164.362602     43.293380      0.817236   
std        1.381587      365.337729     12.574584      0.548138   
min       -1.000000        0.000000      0.000000      0.000000   
25%        0.000000        1.000000     33.000000      1.000000   
50%        0.000000        4.000000     42.000000      1.000000   
75%        1.000000       13.000000     53.000000      1.000000   
max       20.000000     1100.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.505128e+05  
std            1.420324      0.272661  1.098972e+05  
min            0.000000      0.000000  0.000000e+00  
25%            0.000000      0.000000  8.861283e+04  
50%            0.000000      0.000000  1.355147e+05  
75%            1.000000      0.000000  1.955436e+05  
max            4.000000      1.000000  2.265604e+06

здесь уже можно увидеть, что кто-то работает уже 1100 лет. Это явно ошибка.

In [14]:
#заменяю регист в колонке 'edication'
df['education'] = df['education'].str.lower()

In [15]:
#делаю количество детей положильным числом 
df['children'] = df['children'].abs()


In [16]:
#считаю сколько детей может быть 
df['children'].unique()

array([ 1,  0,  3,  2,  4, 20,  5])

In [17]:
#смотрю как распределяется количество детей 
df['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

Чем больше детей, тем меньше человек. Это значит, что 20 детей это ошибка. Думаю, 20 можно изменить на 2.

In [18]:
#заменяю 20 на 2 
df['children'] = df['children'].replace(20, 2)

In [19]:
df['children'].unique()

array([1, 0, 3, 2, 4, 5])

Мне кажется, сейчас значения выглядит куда логичнее.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена верно. 

</div>

### Обработка дубликатов

In [20]:
#смотрю количество дублей
df.duplicated().sum()

71

In [21]:
#вывожу дубликаты
df = df.drop_duplicates().reset_index(drop=True)


In [22]:
#еще раз проверяю
df.duplicated().sum()

0

### Вывод

Больше дубликатов нет, я использовал этот метод, потому что нам нужно было найти дубликаты во всей таблице, а не только в отдельном столбце.
Возможно, дубликаты появились из-за того, что у человека может быть несколько кредитов. Или из-за ошибок на сервере.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно. Скорее всего мы имеем дело с ошибкой. Дубликаты, в основном, удаляются даже если причина их неясна. 

</div>

### Лемматизация

In [23]:
#узнаю количество уникальных значений 
unique_purposes = df['purpose'].unique().tolist()
unique_purposes

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

In [24]:

m = Mystem()
string = ', '.join(unique_purposes)
pprint.pprint(string)

('покупка жилья, приобретение автомобиля, дополнительное образование, сыграть '
 'свадьбу, операции с жильем, образование, на проведение свадьбы, покупка '
 'жилья для семьи, покупка недвижимости, покупка коммерческой недвижимости, '
 'покупка жилой недвижимости, строительство собственной недвижимости, '
 'недвижимость, строительство недвижимости, на покупку подержанного '
 'автомобиля, на покупку своего автомобиля, операции с коммерческой '
 'недвижимостью, строительство жилой недвижимости, жилье, операции со своей '
 'недвижимостью, автомобили, заняться образованием, сделка с подержанным '
 'автомобилем, получение образования, автомобиль, свадьба, получение '
 'дополнительного образования, покупка своего жилья, операции с недвижимостью, '
 'получение высшего образования, свой автомобиль, сделка с автомобилем, '
 'профильное образование, высшее образование, покупка жилья для сдачи, на '
 'покупку автомобиля, ремонт жилью, заняться высшим образованием')


In [25]:
#смотрю сколько слов записано в разных формах
lemmas = m.lemmatize(string)
pprint.pprint(Counter(lemmas))

Counter({' ': 59,
         ', ': 37,
         'покупка': 10,
         'недвижимость': 10,
         'автомобиль': 9,
         'образование': 9,
         'жилье': 7,
         'с': 5,
         'операция': 4,
         'на': 4,
         'свой': 4,
         'свадьба': 3,
         'строительство': 3,
         'получение': 3,
         'высокий': 3,
         'дополнительный': 2,
         'для': 2,
         'коммерческий': 2,
         'жилой': 2,
         'подержать': 2,
         'заниматься': 2,
         'сделка': 2,
         'приобретение': 1,
         'сыграть': 1,
         'проведение': 1,
         'семья': 1,
         'собственный': 1,
         'со': 1,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1,
         '\n': 1})


Можно выделить несколько категорий целей:
Недвижимость 
Автомобиль 
Свадьба
Образование 

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Круто, ты сделал самую сложную часть проекта. Выбор лемм обоснован при помощи подсчета уникальных  слов в параметре целей. 

</div>

### Категоризация данных

In [26]:
def age_group(age):
        if age < 18:
                return 'дети'
        if age <= 30:
                return 'молодые'
        if age <= 64:
                return 'взрослые'
        return 'пенсионеры' 
df['age_group'] = df['dob_years'].apply(age_group)

In [27]:
display(df)


children  years_employed  dob_years education  education_id  \
0             1              23         42    высшее             0   
1             1              11         36   среднее             1   
2             0              15         33   среднее             1   
3             3              11         32   среднее             1   
4             0             932         53   среднее             1   
...         ...             ...        ...       ...           ...   
21449         1              12         43   среднее             1   
21450         0             942         67   среднее             1   
21451         1               5         38   среднее             1   
21452         3               8         38   среднее             1   
21453         2               5         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21449  гражданский брак                 1      F   компаньон     0   
21450   женат / замужем                 0      F   пенсионер     0   
21451  гражданский брак                 1      M   сотрудник     1   
21452   женат / замужем                 0      M   сотрудник     1   
21453   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose   age_group  
0      253875.639453                 покупка жилья    взрослые  
1      112080.014102       приобретение автомобиля    взрослые  
2      145885.952297                 покупка жилья    взрослые  
3      267628.550329    дополнительное образование    взрослые  
4      158616.077870               сыграть свадьбу    взрослые  
...              ...                           ...         ...  
21449  224791.862382             операции с жильем    взрослые  
21450  155999.806512          сделка с автомобилем  пенсионеры  
21451   89672.561153                  недвижимость    взрослые  
21452  244093.050500  на покупку своего автомобиля    взрослые  
21453   82047.418899         на покупку автомобиля    взрослые  

[21454 rows x 13 columns]

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Таблицы на печать выводи с помощью метода display. print переводит данные в строку, а нам этого не надо. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

👍 

</div>

In [28]:
df['age_group'].value_counts()

взрослые      16741
молодые        3717
пенсионеры      895
дети            101
Name: age_group, dtype: int64

Среди клиентов больше всего взрослых людей. В 4 раза больше чем молодых и 17 раз больше чем пенсионеров.
Я категорезировал по возрасту, потому что интересно посмотреть, какая категоря людей берет кредит больше остальных.

In [ ]:
Я не совсем понимаю, что от меня требуется.

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

В данных есть словари. Попробуй их обнаружить и выделить. Это довольно распространенный способ хранения информации. Мне важно понять, что он тобой освоен. Заметь, что в основной таблице мы имеем по 2 параметра с образованием и семейным статусом. Именно по ним и стоит выделить словари. 

</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера 2</h2>

На самом деле словарь ты выделил при помощи Counter выше. Это также словарь, один  из способов хранения информации. Но в задании подразумеваются иные словари. Смотри, в основной таблице по 2 параметра отвечают за семейное положение и уровни образования. Эти параметры следует выделить в отдельные таблицы. Далее мы можем убрать текстовые параметры из основной таблицы, что увеличит эффективность работы с данными. Поскольку это неосновной шаг работы, то оставляю ланное выделение тебе на самостоятельную работу. 

</div>

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [29]:
family_status_set = df.groupby('family_status')['debt'].sum() / \
      df.groupby('family_status')['debt'].count() * 100

In [30]:
family_status_set

family_status
Не женат / не замужем    9.750890
в разводе                7.112971
вдовец / вдова           6.569343
гражданский брак         9.347145
женат / замужем          7.545182
Name: debt, dtype: float64

In [31]:
family_status_pivot = df.pivot_table(index=['family_status'], columns=['debt'], values='total_income', aggfunc='count')
family_status_pivot['ratio'] = family_status_pivot[1] / family_status_pivot[0] * 100
family_status_pivot

debt                       0    1      ratio
family_status                               
Не женат / не замужем   2536  274  10.804416
в разводе               1110   85   7.657658
вдовец / вдова           896   63   7.031250
гражданский брак        3763  388  10.310922
женат / замужем        11408  931   8.160940

Семейное положение влияет. Лучше всех отдают люди в разводе и вдовцы и вдовы, а вот не женатые и не замужем справляются хуже всех.

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

1 означает наличие долга, 0 - его отсутствие. Наименьшую долю должников имеют как раз разведенные и водвцы. Твой вывод не согласуется с полученными результатами. Пожалуйста, исправь это. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Да, теперь все верно. 

</div>

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [32]:
children_set = df.groupby('children')['debt'].sum() / \
    df.groupby('children')['debt'].count() * 100

In [33]:
children_set

children
0    7.543822
1    9.165808
2    9.492481
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

In [34]:
grouped_children = df.groupby('children')['debt'].count()
grouped_children

children
0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: debt, dtype: int64

In [35]:
children_pivot = df.pivot_table(index = ['children'], \
                                columns = ['debt'], \
                                values = 'purpose', aggfunc='count')

children_pivot['ratio'] = children_pivot[1] / children_pivot[0] * 100
children_pivot

debt            0       1      ratio
children                            
0         13028.0  1063.0   8.159349
1          4410.0   445.0  10.090703
2          1926.0   202.0  10.488058
3           303.0    27.0   8.910891
4            37.0     4.0  10.810811
5             9.0     NaN        NaN

Наличие детей на возврат кредита влияют. Бездетные отдают кредиты лучше.

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Также подкорректируй выводы. 

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [36]:
df['total_income'].quantile([0.25,0.5,0.75])

0.25     89089.043938
0.50    135781.042522
0.75    195813.983455
Name: total_income, dtype: float64

In [37]:
def income_status(total_income):
    if total_income <= 88946.5:
            return 'Низкий уровень дохода'
    if total_income <= 135716.0:
            return 'Средний уровень дохода'
    if total_income < 195751.5:
            return 'Высокий уровень дохода'
    return 'Сверхвысокий уровень дохода'

In [38]:
df['income_status'] = df['total_income'].apply(income_status)

In [39]:
df.head()

children  years_employed  dob_years education  education_id  \
0         1              23         42    высшее             0   
1         1              11         36   среднее             1   
2         0              15         33   среднее             1   
3         3              11         32   среднее             1   
4         0             932         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose age_group                income_status  
0               покупка жилья  взрослые  Сверхвысокий уровень дохода  
1     приобретение автомобиля  взрослые       Средний уровень дохода  
2               покупка жилья  взрослые       Высокий уровень дохода  
3  дополнительное образование  взрослые  Сверхвысокий уровень дохода  
4             сыграть свадьбу  взрослые       Высокий уровень дохода

In [40]:
total_income_set = df.groupby('income_status')['debt'].sum() / \
      df.groupby('income_status')['debt'].count() * 100

In [41]:
total_income_set.sort_values()

income_status
Сверхвысокий уровень дохода    7.134873
Низкий уровень дохода          7.830312
Средний уровень дохода         8.477734
Высокий уровень дохода         9.016393
Name: debt, dtype: float64

In [42]:
total_income_pivot = df.pivot_table(index=['income_status'], columns=['debt'], values='education_id', aggfunc='count')
total_income_pivot['ratio'] = total_income_pivot[1] / total_income_pivot[0] * 100
total_income_pivot

debt                            0    1     ratio
income_status                                   
Высокий уровень дохода       4884  484  9.909910
Низкий уровень дохода        4932  419  8.495539
Сверхвысокий уровень дохода  4985  383  7.683049
Средний уровень дохода       4912  455  9.263029

In [43]:
#проверяю количество людей в каждой из групп
df['income_status'].value_counts(normalize = True)

Высокий уровень дохода         0.250210
Сверхвысокий уровень дохода    0.250210
Средний уровень дохода         0.250163
Низкий уровень дохода          0.249417
Name: income_status, dtype: float64

Возвращаймость кредита от уровня дохода есть. Лучше всего возвращают люди со свердоходами, а высоким хуже всех. Тоже результат неожиданный.

<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Здорово, что ты разбил клиентов на равные группы. Также можно использовать разбиение по квантилям данных при помози метода qcut. Тогда группы также получатся одинакового размера. Или же можно узнать в интернете процентный состав общества по доходам и затем уже категоризовать столбец с доходами по найденному соотношению с помощью персентилей. 

</div>

- Как разные цели кредита влияют на его возврат в срок?

In [46]:
purpose_keys= []
purpose_list = list(df.purpose.unique())
for value in range(len(purpose_list)):
    if 'образован' in purpose_list[value]:
        purpose_keys.append('Образование')
    elif 'авто' in purpose_list[value]:
        purpose_keys.append('Автомобиль')
    elif 'свадь' in purpose_list[value]:
        purpose_keys.append('Свадьба')
    else:
        purpose_keys.append('Недвижимость')
        
purpose_keys

['Недвижимость',
 'Автомобиль',
 'Образование',
 'Свадьба',
 'Недвижимость',
 'Образование',
 'Свадьба',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Автомобиль',
 'Автомобиль',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Недвижимость',
 'Автомобиль',
 'Образование',
 'Автомобиль',
 'Образование',
 'Автомобиль',
 'Свадьба',
 'Образование',
 'Недвижимость',
 'Недвижимость',
 'Образование',
 'Автомобиль',
 'Автомобиль',
 'Образование',
 'Образование',
 'Недвижимость',
 'Автомобиль',
 'Недвижимость',
 'Образование']

In [47]:
purpose_list

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

In [48]:
purpose_dict = dict(zip(purpose_list, purpose_keys))

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Это также является словарем. Главное, что ты знаешь об этом способе хранения информации. 

</div>

In [49]:
df['purpose'] = df['purpose'].map(purpose_dict)

In [50]:
purpose_pivot = df.pivot_table(index=['purpose'], columns=['debt'], values='education_id', aggfunc='count')
purpose_pivot['ratio'] = purpose_pivot[1] / purpose_pivot[0] * 100
purpose_pivot

debt              0    1      ratio
purpose                            
Автомобиль     3903  403  10.325391
Недвижимость  10029  782   7.797388
Образование    3643  370  10.156464
Свадьба        2138  186   8.699719

Цели кредита влияют. Лучше всего отдают кредиты на свадьбу и на недвижимость. На автомобиль и на образование хуже. Для меня результаты неожиданные.


<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Здорово, что при анализе использован метод сводных таблиц. Он пригодится в дальнейших проектах.

Выводы сделаны по всем пунктам. Однако по всем же пунктам пройди и подкорректируй их, т.к. они неверны. Выводы должны быть обратны, т.к. ты не совсем верно трактуешь 0 и 1 в столбце с долгом. 

</div>

### Шаг 4. Общий вывод

Общий вывод следующий:
a) Я поработал над ошибками в таблице: заменил пропуски на медианные показатели по группе, удалил дубликаты, заменил количество детей с 20 на 2, привел столбец оброзавание к одному регистру, перевел стаж в года. 
b) Я выяснил 4 основных цели получения кредита это: автомобиль, недвижимость, свадьба, образование. У всех они сформулированы по-разному.
c) Я выяснил, как влиет доход, цель кредита, наличие детей и семейное положение на возврат кредита. Самыми ответсвенными оказались люди берущие кредит на недвижимость, бездетные, состоящие в браке и высоким уровнем дохода. Самыми безответсвенными являются люди берущие кредит на машину, с детьми, не состаящие в браке и среднем уровнем дохода.


<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Финальный вывод и есть главный результат твоей работы. Стоит писать его подробно по результатам проведенной работы. В нем можно приводить полученные в ходе работы значения. Также можно расписать все, что было сделано в работе.

</div>

<div class="alert alert-danger">
<h2> Комментарий ревьюера 2</h2>

Кажется, твой финальный вывод не сохранился. Пожалуйста, допиши его. Удели основное внимание полученным в ходе работы результатам. 

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения. Здорово, что по каждому пункту работы приведен вывод. Старайся делать больший акцент именно на полученных в работе результатах.

</div>

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

- В начале работы следует описывать данные, с которыми работаешь;
- Попробуй улучшить обработку пропусков;
- Выдели словари в данных;
- Исправь выводы по финальному пункту анализа;
- Финальный вывод стоит писать подробно.

Ты проделал большую работу, молодец! Исправь отмеченные выше помарки. Также есть достаточное число желтых комментариев. Думаю, исправить их тебе не составит труда. Жду твою работу :)

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 2</h2>

Ты поработал со всеми помарками, молодец. Осталось лишь дописать финальный вывод. Тогда работа будет выполнена полностью. Как сделаешь это, присылай проект снова. Буду ждать :)

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера 3</h2>

Помарки исправлены, и теперь работа выполнена хорошо. Ты здорово поработал, молодец. Спасибо за оставленные комментарии по исправлениям. Успехов в дальнейших проектах :)

</div>
   

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.